src/raftstore/store/snap.rs

```diff
@@ -618,6 +620,987 @@ mod v1 {
+                    Some(x) => {
+                        let expected_size = x.1;
+                        let expected_checksum = x.2;
+                        let size = try!(get_file_size(&cf_file.path));
+                        if size != expected_size {
+                            return Err(box_err!("invalid snapshot file size {} for cf {}, \
+                                                 expected {}",
+                                                size,
+                                                cf_file.cf,
+                                                expected_size));
+                        }
+                        let checksum = try!(calc_checksum(&cf_file.path));
+                        if checksum != expected_checksum {
+                            return Err(box_err!("invalid snapshot file checksum {} for cf {}, \
+                                                 expected {}",
+                                                checksum,
+                                                cf_file.cf,
+                                                expected_checksum));
+                        }
+                    }
+                    None => {
+                        return Err(box_err!("missing snapshot file size and checksum for cf {}",
+                                            cf_file.cf));
+                    }
+                }
+            }
+            Ok(())
+        }
+
+        fn switch_to_cf_file(&mut self, cf: &str) -> io::Result<()> {
+            match self.cf_files.iter().position(|x| x.cf == cf) {
+                Some(index) => {
+                    self.cf_index = index;
+                    Ok(())
+                }
+                None => Err(io::Error::new(ErrorKind::Other, format!("fail to find cf {}", cf))),
+            }
+        }
+
+        fn add_kv(&mut self, k: &[u8], v: &[u8]) -> io::Result<()> {
+            let mut cf_file = &mut self.cf_files[self.cf_index];
+            let mut writer = cf_file.sst_writer.as_mut().unwrap();
+            if let Err(e) = writer.add(k, v) {
+                return Err(io::Error::new(ErrorKind::Other, e));
+            }
+            cf_file.kv_count += 1;
+            Ok(())
+        }
+
+        fn save_cf_files(&mut self) -> io::Result<()> {
+            for cf_file in &mut self.cf_files {
+                if cf_file.kv_count == 0 {
+                    let _ = cf_file.sst_writer.take().unwrap();
+                } else {
+                    let mut writer = cf_file.sst_writer.take().unwrap();
+                    if let Err(e) = writer.finish() {
+                        return Err(io::Error::new(ErrorKind::Other, e));
+                    }
+                }
+                try!(fs::rename(&cf_file.tmp_path, &cf_file.path));
+                cf_file.size = try!(get_file_size(&cf_file.path));
+                cf_file.checksum = try!(calc_checksum(&cf_file.path));
```

Comment:
It can't calculate the checksum when adding kv into a `SstFileWriter`, because it's unknow what's written into the sst file exactly before finishing a `SstFileWriter`.